File: Java/Lesson085_ConstructBinaryTreeFromPreorderAndInorder.java
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 085 -- Construct Binary Tree from Preorder and Inorder
// Category   : Trees
// Difficulty : Medium
// Study Plan : Day 43
// =============================================================
//
// QUESTION:
//   Given preorder and inorder traversals of a binary tree (no duplicates), construct and return the tree.
// =============================================================

import java.util.*;
public class Lesson085_ConstructBinaryTreeFromPreorderAndInorder {
    static class TreeNode { int val; TreeNode left, right; TreeNode(int v){val=v;} }
    int p = 0; Map<Integer,Integer> idx = new HashMap<>();
    TreeNode go(int[] preorder, int l, int r){
        if (l > r) return null;
        int v = preorder[p++]; TreeNode root = new TreeNode(v); int m = idx.get(v);
        root.left = go(preorder, l, m-1); root.right = go(preorder, m+1, r);
        return root;
    }
    public TreeNode buildTree(int[] preorder, int[] inorder){
        for (int i=0;i<inorder.length;i++) idx.put(inorder[i], i);
        return go(preorder, 0, inorder.length-1);
    }
    static List<Integer> pre(TreeNode n){ List<Integer> r=new ArrayList<>(); if (n==null) return r;
        r.add(n.val); r.addAll(pre(n.left)); r.addAll(pre(n.right)); return r; }
    public static void main(String[] a){
        TreeNode t = new Lesson085_ConstructBinaryTreeFromPreorderAndInorder().buildTree(new int[]{3,9,20,15,7}, new int[]{9,3,15,20,7});
        System.out.println(pre(t));
    }
}

File: Go/Lesson_085_construct_binary_tree_from_preorder_and_inorder.go
//go:build ignore

// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 085 -- Construct Binary Tree from Preorder and Inorder
// Category   : Trees
// Difficulty : Medium
// Study Plan : Day 43
// =============================================================
//
// QUESTION:
//   Given preorder and inorder traversals of a binary tree (no duplicates), construct and return the tree.
// =============================================================

package main
import "fmt"
type TreeNode struct { Val int; Left, Right *TreeNode }
func buildTree(preorder, inorder []int) *TreeNode {
    idx := map[int]int{}
    for i,v := range inorder { idx[v] = i }
    p := 0
    var go func(l, r int) *TreeNode
    go = func(l, r int) *TreeNode {
        if l > r { return nil }
        v := preorder[p]; p++
        root := &TreeNode{Val: v}; m := idx[v]
        root.Left = go(l, m-1); root.Right = go(m+1, r)
        return root
    }
    return go(0, len(inorder)-1)
}
func pre(n *TreeNode, out *[]int){ if n == nil { return }; *out = append(*out, n.Val); pre(n.Left, out); pre(n.Right, out) }
func main(){ t := buildTree([]int{3,9,20,15,7}, []int{9,3,15,20,7})
    out := []int{}; pre(t, &out); fmt.Println(out) }
